README
##################################################################
CREDITS
##################################################################

phpFB is a simple and easy to use PHP script which lists the content
of web folders. It is designed to look similar to common web servers.
Easily customizable and tweakable.

Copyright (C) 2006, Berke Viktor

##################################################################
THINGS YOU NEED TO KNOW
##################################################################

First of all: the way phpFB reports things is currently being
revised, so don't be surpised about arrays with one element etc :-)

------------------------------------------------------------------

The script itself is pretty simple IMHO, you can tweak it to your
needs, but I placed some of the important settings in the 
beginning of the script to make your life easier (see "OPTIONS").

------------------------------------------------------------------

To 'install' phpFB, you have nothing else to do but place it in a
folder which you want to show off to the public.

------------------------------------------------------------------

You shouldn't have folders with ".." in their names. For example:

PHP is...fun

You probably don't want such folders, but it's just better to make
clear. If you asked why, I would say this is needed to disable 
moving out the "root" directory (the directory which phpFB is 
placed in).

------------------------------------------------------------------

File uploads need a temporary directory. You can modify this only
if you have access to the host's config files. Hopefully you don't
have to care about this, but if uploads don't work for you, it's
your task to find out what's this dir. You might query it by
running phpinfo(), but it could be disabled on some hosts.

------------------------------------------------------------------

You can specify upload sizelimit in two ways. The first one is 
$up_size in phpFB's options section. The second one is to specify
it in @UPLOAD in bytes. This overrides the default value only to
its folder. Warning! If the file contains only spaces, you'll have
0 MB sizelimit in that folder. An example @UPLOAD could contain:

1024

Which sets sizelimit to 1 kB.

------------------------------------------------------------------

If the length of a file's name is more than 50 characters, it's
ending will be chunked till the 47th character, and "..." will be
placed in the end of it. Note that this applies to listing only,
no file names will be altered.

------------------------------------------------------------------

I setted up cells in the table to look acceptable both in Firefox
and Epiphany on Linux. If you have problems with viewing, please
let me know.

------------------------------------------------------------------

If you want your file to be hidden, start it with @. You could use
a point (.) too, but this is not recommended, it can cause problems,
especially on free web spaces.

------------------------------------------------------------------

Files ending to .mp3, .wma are not uploadable by default. It could
be disabled easily, look for "mp3" or "wma" and you'll see ;-).

------------------------------------------------------------------

PHP files are uploadable, but they will be renamed to .phps to
avoid hacks.

------------------------------------------------------------------

The uploader is not allowed to overwrite existing files. If you
want to upload something which has the same name, you have to
rename it.

##################################################################
IMPORTANT FILES THAT MODIFY FUNCTIONALIY
##################################################################

There are files which you can use to control the behaviour of the
script in each folder. Besides @HEADER they could be empty files.
(Actually @HEADER could be empty, too, but the use of would be
quite nonsense then.)

------------------------------------------------------------------

@VIEWER

Place a file called @VIEWER into the folder which you want to 
enable viewer in.

------------------------------------------------------------------

@HEADER

Place a file called @HEADER into the folder you want to have a note
displayed in top of it. It can contain more lines and HTML tags, too.

------------------------------------------------------------------

@PRIVATE

Place a file called @PRIVATE into the folder you want to hide from
visitors. Warning, this is not recursive, it applies to that single
folder only, subfolders are not affected.

------------------------------------------------------------------

@UPLOAD

Place a file called @UPLOAD into the folder which you want to
enable uploads in. This file can also be used to specify custom
sizelimits in any folder. This function was mentioned previously.

##################################################################
OPTIONS
##################################################################

$gfxdir = '@gfx';

This folder contains the necessairy icons which are required to
display the content properly. Enter the URL or path of your gfx
folder without ending slash. Both kind is accepted, examples:

$gfxdir = 'pub/files/gfx_dir';
$gfxdir = 'http://example.com/gfx';

------------------------------------------------------------------

$up_size = '2097152';

Here you can specify the maximum size of files (in bytes) that
one can upload using the uploader.

------------------------------------------------------------------

$chmod_enabled = '1';

This must be enabled to grant access to files uploaded via 
phpFB since some hosts don't provide access to them by default. 
If you know that this is needless on your host or the chmod()
function is disabled, comment it out with // or change "1" to 
anything you like.

##################################################################


CHANGELOG

//////////////////////
UNDONE:
//////////////////////

- Reload page after upload.

- Rewrite 'messaging system'.

//////////////////////
DONE:
//////////////////////

- Fix open_basedir restriction errors on certain hosts.
(1.0.1)

- Insert returning messages into popup warnings.
(1.0.0)

- Supply missing quotation marks (HTML).
(1.0.0)

- Uploader is also valid HTML now.
(1.0.0)

- Overwrite protection added to uploader.
(1.0.0)

- Icon alternative text replaced.
(1.0.0)

- Uploader revised. Changes include php file upload as phps.
(1.0.0)

- Uploader: file ext check works correctly both with up and
lowcase files.
(1.0.0)

- Uploader: file extension check instead of file type check.
(1.0.0)

- @HEADER files now can contain more lines and HTML tags.
(1.0.0)

- Files starting with '.' (UNIX hidden files) are hidden now.
(1.0.0)

- phpFB hid all files containing @, fixed.
(1.0.0)

- Fix problem about trying to get filemtime of '..'
(1.0.0)

- Define max upload size per folder by @UPLOAD.
(0.9.9)

- Rename files longer than 50 characters.
(0.9.9)

- Solve problem with empty files.
(0.9.8)

- Fix trim problems.
(0.9.8)

- Make phpFB valid HTML 4.01 (except with uploader).
(0.9.8)

- Display 'file last modified' info.
(0.9.8)

- Replace global switches of gallery/upload with file checks.
(0.9.7)

- Introducing private folders.
(0.9.6)

- Introducing header function.
(0.9.6)

- "chroot" to dir only if it's the own dir.
(0.9.5)

- Fix problem with spaces in filenames.
(0.9.5)

- Uploader recovered.
(0.9.5)

- Problem with phpFB placed in docroot fixed.
(0.9.5)

- Unify lister and gallery script.
(0.9.4)

- Gallery extension recovered.
(0.9.4)

- New option to set cell width containing filenames.
(0.9.3)

- Some cleanup.
(0.9.3)

- Removal of some unused variables.
(0.9.3)

- Rewrite of file checks in a tidier and simpler form.
(0.9.3)

- Ban of php uploads.
(0.9.2)

- Chroot-like stuff.
(0.9.2)

- Recover most of the functionality that were available in the 
per-folder version.
(0.9.2)

- Huge changes, phpFB is one simple file instead of one file 
per folder.
(0.9.1)

//////////////////////
REMOVED/SUBSTITUTED:
//////////////////////

- Associate file icons by MIME type
http://hu2.php.net/manual/hu/function.mime-content-type.php
Reason: the required function is disabled on most hosts.
Only 3 type remains: folders, image files, other files.

- Display size not only of files but of folders, too
Reason: script execution time would probably get very high
especially when there were many folders, and also would 
probably cause high CPU/HDD load on the host.

- Check for uname
Reason: this would be needed for win hosts. But I realized that
I don't care about win :-). So if you need phpFB on win, you'll
have to fix it yourself, sorry.

- Make gallery look better.
Reason: I can't imagine a better scheme :-).


lister.php
<?php
/***************************************************************************

   Copyright (c) 2006-2009, Berke Viktor
   All rights reserved.
   
   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions
   are met:
   1. Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
   2. Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
   3. The name of the author may not be used to endorse or promote products
      derived from this software without specific prior written permission.
   
   THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
   IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
   OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
   IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
   INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
   NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
   THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

***************************************************************************/

/* options */
$gfxdir = '@gfx';
$up_size = '2097152';
$chmod_enabled = '1';
/* end of options */

/* bootstrap */
$phpfb_ver = 'trunk';
$wdir = getcwd();

$dir1 = ltrim($_GET[dir], "/");

$messages = array(
	"perm_error" => "You are not allowed to browse the selected directory."
);

/* set doctype of output */
echo '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">';

$dir2 = str_replace(" ", "%20", $dir1);

$back1 = explode("/", $dir1);
$back2 = ((count($back1)) - 2);

$i = 0;
while ($back2 > $i)
{
	$back3 = $back3.'/'.$back1[$i];
	$i++;
}

/* display header file if it exists */
if (file_exists($dir1.'@HEADER'))
{
	$header = file_get_contents($dir1.'@HEADER');
}

/* if folder is set to private, return to parent folder */
if (file_exists($dir1.'@PRIVATE') or strpos($dir1, '..') > -1)
{
	echo '<head><meta http-equiv="refresh" content="0,url=lister.php?priv=1&dir='.$back3.'" /></head></html>';
	die;
}

if ($_GET[priv] == 1)
{
	$message = $messages["perm_error"];
}
/* end of bootstrap */

/* start of functions */
function do_upload($upload_dir, $upload_url, $up_size, $chmod_enabled)
{
	$temp_name = $_FILES['userfile']['tmp_name'];
	$file_name = $_FILES['userfile']['name']; 
	$result    = $_FILES['userfile']['error'];

	$file_name2 = explode('.', $file_name);
	$file_ext = strtoupper($file_name2[(count($file_name2) - 1)]);

	if (file_exists($upload_dir.$file_name))
	{
		$message = 'The file already exists!';
		return $message;
	}
	if ($_FILES['userfile']['size'] > $up_size)
	{
		$message = 'The file is too big!';
		return $message;
	}
	if ($file_ext == 'MP3' or $file_ext == 'WMA')
	{
		$message = $file_ext.' upload is disabled.';
		return $message;
	}

	if ($chmod_enabled == '1')
	{
		chmod($temp_name, 0644);
	}

	if ($file_ext == 'PHP')
	{
		$result = move_uploaded_file($temp_name, $upload_dir.$file_name.'s');
	}
	else
	{
		$result = move_uploaded_file($temp_name, $upload_dir.$file_name);
	}

	$message = ($result)?'Success! Click on the folder name to see your file!' : 'Upload failed!';
	return $message;
}
/* end of functions */

/* start of viewer */
if ($_GET[view] == 1)
{
	$img1 = $_GET[fn];
	$img2 = str_replace(" ", "%20", $img1);
	$pic = explode("/", $img2);
	$picname2 = $pic[(count($pic))-1];
	$picname1 = str_replace("%20", " ", $picname2);
	$pic_back = rtrim($img2, "$picname2");
	$reso = getimagesize($img1);
	$res_w = $reso[0];
	$res_h = $reso[1];
	
	/* some grammar */
	$size_str = (string)number_format(filesize($img1)/1024, 0);
	if ($size_str > 1)
	{
		$size_str .= " kBytes";
	}
	else
	{
		$size_str .= " kByte";
	}

	/* output the result */
	echo '<head><meta http-equiv="content-type" content="application/xhtml+xml; charset=utf-8" />
	<title>'.$picname1.'</title></head><body link="#0000cc" alink="#0000cc" vlink="#0000cc"><div>
	<img alt="back" src="'.$gfxdir.'/back.png" border="0" /><a href="lister.php?dir='.$pic_back.'">Back</a><br /><br />
	<b>Filename: </b><a href="'.$img2.'">'.$picname1.'</a><br />
	<b>Filesize: </b>'.$size_str.'<br />
	<b>Resolution: </b>'.$res_w.' x '.$res_h.' pixels<br />
	<hr width="'.($reso[0] + 2).'" align="left" />
	<img alt="image" src="'.$img2.'" width="'.$res_w.'" height="'.$res_h.'" border="1" />
	<hr width="'.($reso[0] + 2).'" align="left" />';
} 
/* end of viewer */

/* start of lister */
else
{
	//  >>>>>  LISTER  >>>>>
	$the_array = Array();
	$handle = opendir($wdir.'/'.$dir1);
	
	while (false !== ($file1 = readdir($handle)))
	{
		$file2 = str_replace(" ", "%20", $file1);
		if ($file2 != $dir1.'lister.php' && substr($file2, 0,1) != '@' && substr($file2, 0,1) != '.' xor substr($file2, 0,2) == '..')
		{
			$the_array[] = $file2;
		}
	}

	closedir($handle);
	sort ($the_array);
	reset ($the_array);
	while (list ($key, $val2) = each ($the_array))
	{
		$val1 = str_replace("%20", " ", $val2);

		if ($val1 == '..' || is_file($dir1.$val1) == false)
		{
			$f_icon = 'folder.png';
			$f_size = '-';
			$f_link = 'lister.php?dir='.$dir2.$val2.'/';
			$f_alt= '[DIR]';
		}
		else
		{
			$f_size = number_format(filesize($dir1.$val1)/1024, 0, ',', ' ').' kB';
			$f_alt= '[ . ]';
			if ($f_size == 0)
			{
				$f_icon = 'unknown.png';
				$f_link = $dir2.$val2;
			}
			else
			{
				$f_check = getimagesize($dir1.$val1);
				if ( $f_check[2] == '' )
				{
					$f_icon = 'unknown.png';
					$f_link = $dir2.$val2;
				}
				else
				{
					$f_icon = 'image.png'; 
					if (file_exists($dir1.'@VIEWER'))
					{
						$f_link = 'lister.php?view=1&amp;fn='.$dir2.$val2;
					}
					else
					{
						$f_link = $dir2.$val2;
					}
				}
			}
		}
		
		if ($val1 != '..')
		{
			$f_date = date("d-m-Y H:i", filemtime($dir1.$val1));
		}

		if ($val1 == '..')
		{
			$f_link = 'lister.php?dir='.$back3.'/';
			$val1 = 'Parent Directory';
			$f_icon = 'back.png'; 
			$f_date ='';
			$f_size = '-';
		}

		if ((strlen($val1)) > 50)
		{
			$val1 = substr($val1, 0, 47).'...';
		}

		$file_list .= '<table border="0" cellspacing="0" cellpadding="0"><tr>
			<td width="500"><tt><img alt="'.$f_alt.'" src="'.$gfxdir.'/'.$f_icon.'" /> <a href="'.$f_link.'">'.$val1.'</a></tt></td>
			<td width="180" align="right"><tt>'.$f_date.'</tt></td>
			<td width="110" align="right"><tt>'.$f_size.'</tt></td>
			</tr></table>';
	}

	echo '<head><meta http-equiv="content-type" content="application/xhtml+xml; charset=utf-8" />
		<title>Index of /'.$dir1.'</title></head><body link="#0000cc" alink="#0000cc" vlink="#0000cc">
		<h1>Index of <a href="lister.php?dir='.$dir2.'" style="text-decoration: none;"><font color="#000000">/'.$dir1.'</font></a></h1>
		<div><table border="0" cellspacing="0" cellpadding="0"><tr>
		<td width="500"><tt><img alt="spacer" src="'.$gfxdir.'/blank.png" /> Name</tt></td>
		<td width="180" align="right"><tt>Last Modified</tt></td>
		<td width="110" align="right"><tt>Size</tt></td>
		</tr></table><hr />';
	if ($header)
	{
		echo '<table border="0" cellspacing="0" cellpadding="0"><tr><td width="500"><tt>'.$header.'</tt></td></tr></table><br />';
	}
	echo $file_list.'<hr />';
/* end of lister */

/* start of uploader */
	if (file_exists($dir1.'@UPLOAD'))
	{
		$up_size2 = file_get_contents($dir1.'@UPLOAD');
		if ($up_size2 != '')
		{
			$up_size = $up_size2;
		}

		$upload_dir = $wdir.'/'.$dir1;
		$upload_url = 'http://'.$_SERVER['HTTP_HOST'].$dir1;

		if ($_FILES['userfile'])
		{
			$message = do_upload($upload_dir, $upload_url, $up_size, $chmod_enabled);
		}

		echo '<form name="upload" enctype="multipart/form-data" method="post" action="">
			<span class="textm">Uploader ('.number_format($up_size/1048576, 2, ',', ' ').' MB): </span>
			<input type="file" id="userfile" name="userfile" />
			<input type="submit" name="upload" value="Upload" /></form><hr />';

	}
/* end of uploader */
}

/* start of footer */
echo '</div><address>phpFB '.$phpfb_ver.' (<a href="http://validator.w3.org/check?uri=referer">validator</a>)</address></body></html>';

if ($message)
{
	echo '<script language="JavaScript">window.alert("'.$message.'");</script>';
}
/* end of footer */
?>
